### PR TITLE
planner/core: code refactor around join reorder

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

some code refactor.

### What is changed and how it works?

Extract common code that dp/greedy will use.

### Check List 

Tests 

 - Unit test(existing ones can cover it).

### CODE REVIEW

File: planner/core/rule_join_reorder.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+type jrNode struct {
+	p       LogicalPlan
+	cumCost float64
+}
+
+func (s *joinReOrderSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
+	return s.optimizeRecursive(p.context(), p)
+}
+
+func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalPlan) (LogicalPlan, error) {
```
Comment (shenli) [inline planner/core/rule_join_reorder.go]: Please add some comments **in** this function and the following functions.

File: planner/core/rule_join_reorder.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	p       LogicalPlan
+	cumCost float64
+}
+
+func (s *joinReOrderSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
+	return s.optimizeRecursive(p.context(), p)
+}
+
+// optimizeRecursive first checks that whether the current plan tree is start with a join group.
+// If so it will call the reorder algorithm to reorder the join group.
```
Comment (zz-jason) [inline planner/core/rule_join_reorder.go]: how about:
```go
// optimizeRecursive recursively collects join groups and applies the greedy
// join reorder algorithm for each group.
```

File: planner/core/rule_join_reorder.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+type baseSingleGroupJoinOrderSolver struct {
+	ctx          sessionctx.Context
+	curJoinGroup []*jrNode
+	otherConds   []expression.Expression
+}
+
+// baseNodeCumCost calculate the cumulative cost of the node in the join group.
+func (s *baseSingleGroupJoinOrderSolver) baseNodeCumCost(groupNode LogicalPlan) float64 {
```
Comment (zz-jason) [inline planner/core/rule_join_reorder.go]: There is already a `base` keyword in `baseSingleGroupJoinOrderSolver`, how about renaming `baseNodeCumCost` to `calcNodeCumCost`?

And I found `baseSingleGroupJoinOrderSolver.calcJoinCumCost()` can be removed. In `joinReorderGreedySingleGroupSolver.constructConnectedJoinTree()`, `newJoin` is constructed with `curJoinTree` and `node`, so we can directly call `baseSingleGroupJoinOrderSolver.baseNodeCumCost(newJoin)` to achieve the same purpose.

Comment (winoros) [reply planner/core/rule_join_reorder.go]: When join tree is large, use `baseNodeCumCost` to search the subtree again and again is not a good idea...

File: planner/core/rule_join_reorder.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+package core
+
+import (
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+// extractJoinGroup extracts all the join nodes connected with continuous
+// InnerJoins to construct a join group. This join group is further used to
+// construct a new join order based on a greedy algorithm.
```
Comment (lamxTyler) [inline planner/core/rule_join_reorder.go]: What about removing the `greedy`? It will also be used by dp.

**Pull request merged**